irobot_events_executor/src/rclcpp/executors/events_executor/events_executor_entities_collector.cpp

```diff
@@ -97,6 +98,18 @@ EventsExecutorEntitiesCollector::execute(std::shared_ptr<void> & data)
 
   std::lock_guard<std::recursive_mutex> lock(reentrant_mutex_);
 
+  // Clear maps to ensure that after execute, they hold non-expired entities.
+  // This fixes the issue of new entities sharing the same memory released by
+  // a previous entity, where trying to lock the weak pointer failed.
+  weak_clients_map_.clear();
+  weak_services_map_.clear();
+  weak_waitables_map_.clear();
+
+  // Copy non-user waitables to weak_waitables_map_
```

Comment:
Can you explain why we need to treat "non user waitables" differently? I'm not understanding this change

Reply:
What I call "user waitables" are the waitables that an user creates, like intra-process subscriptions or actions, and they are retrieved from `groups` here:
https://github.com/irobot-ros/events-executor/blob/30d99c4e0c0dd96d05dbf14d5a00da162c6d040f/irobot_events_executor/src/rclcpp/executors/events_executor/events_executor_entities_collector.cpp#L251-L260

The "non-user waitables" are the `EventsExecutorNotifyWaitable`:
https://github.com/irobot-ros/events-executor/blob/93e58873d02cd3b41f7459f0fcbbf1487807b9eb/irobot_events_executor/src/rclcpp/executors/events_executor/events_executor.cpp#L46
and the `EventsExecutorEntitiesCollector`:
https://github.com/irobot-ros/events-executor/blob/30d99c4e0c0dd96d05dbf14d5a00da162c6d040f/irobot_events_executor/src/rclcpp/executors/events_executor/events_executor_entities_collector.cpp#L88

The "non-user" waitables are only created by the executor, so no risk of being destroyed and other waitables reusing their memory addresses (I could call them "executor waitables").

The "user" waitables can be destroyed and their memory reused, so I'm interested only in clearing and reseting those ones, to have always the "latest" weak pointer.

So at every `execute()` iteration of the `EventsExecutorEntitiesCollector`, I clear the map `weak_waitables_map_` but right away I re-add the non-user waitables (which are always valid) and stored in `non_user_weak_waitables_map_`. 
The "user" waitables are re-added but using their latest "valid" weak pointers.